File: app/Modules/Gateway/Http/Controllers/ActivationStatusController.php
<?php

namespace App\Modules\Gateway\Http\Controllers;

use App\Modules\Gateway\Models\Transaction;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Http;

/**
 * Tokenowany endpoint dla zewnętrznego monitoringu aktywacji konta PayU.
 * Sygnał aktywacji: PayU provisionuje metody płatności na nowym POS-ie
 * (4433543, sklep zweryfikowany pod https://please-support-me.com).
 */
class ActivationStatusController extends Controller
{
    public function show(Request $request)
    {
        if (! hash_equals((string) config('payment.activation_check_token'), (string) $request->query('token'))) {
            return response()->json(['error' => 'unauthorized'], 401);
        }

        $newPos = ['methods' => null, 'blik_enabled' => null, 'error' => null];

        try {
            $auth = Http::asForm()->timeout(10)->post('https://secure.payu.com/pl/standard/user/oauth/authorize', [
                'grant_type' => 'client_credentials',
                'client_id' => config('payment.payu_newpos.client_id'),
                'client_secret' => config('payment.payu_newpos.client_secret'),
            ]);
            $token = $auth->json('access_token');

            if ($token) {
                $methods = Http::withToken($token)->acceptJson()->timeout(10)
                    ->get('https://secure.payu.com/api/v2_1/paymethods', ['lang' => 'pl'])
                    ->json('payByLinks', []);

                $enabled = array_values(array_filter($methods, fn ($m) => ($m['status'] ?? '') === 'ENABLED'));
                $newPos['methods'] = count($enabled);
                $newPos['blik_enabled'] = collect($enabled)->contains(fn ($m) => $m['value'] === 'blik');
            } else {
                $newPos['error'] = 'oauth_failed';
            }
        } catch (\Throwable $e) {
            $newPos['error'] = 'exception';
        }

        return response()->json([
            'checked_at' => now()->toIso8601String(),
            'new_pos' => $newPos,
            // czy realne płatności się księgują (przelewy z banków docierają)
            'paid_last_24h' => Transaction::where('status', 'paid')
                ->where('paid_at', '>', now()->subDay())->count(),
            'pending_last_24h' => Transaction::where('status', 'pending')
                ->where('created_at', '>', now()->subDay())->count(),
            'activation_likely_done' => ($newPos['methods'] ?? 0) > 0,
        ]);
    }
}

File: pay-backend/gateway-api/src/main/kotlin/com/supportme/pay/gateway/api/ops/ActivationStatusController.kt
package com.supportme.pay.gateway.api.ops

import com.fasterxml.jackson.databind.ObjectMapper
import com.supportme.pay.gateway.api.config.PayUNewPosConfigProperties
import com.supportme.pay.gateway.api.config.PaymentConfigProperties
import com.supportme.pay.gateway.domain.entity.TransactionStatus
import com.supportme.pay.gateway.domain.repository.TransactionRepository
import com.supportme.pay.gateway.payments.PayUOAuthTokenCache
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.client.RestClient
import java.security.MessageDigest
import java.time.Instant
import java.time.temporal.ChronoUnit

data class NewPosStatus(val methods: Int? = null, val blikEnabled: Boolean? = null, val error: String? = null)
data class ActivationStatusResponse(
    val checkedAt: String,
    val newPos: NewPosStatus,
    val paidLast24h: Long,
    val pendingLast24h: Long,
    val activationLikelyDone: Boolean,
)

/**
 * Odpowiednik `ActivationStatusController` — tokenowany endpoint dla
 * zewnętrznego monitoringu aktywacji konta PayU. Sygnał aktywacji: PayU
 * provisionuje metody płatności na nowym POS-ie. ZAWSZE bije w PRODUKCYJNE
 * `secure.payu.com`, niezależnie od `payu.env` (sandbox/production) —
 * to jest osobny, docelowo produkcyjny POS, nie ten obsługujący ruch.
 */
@RestController
@RequestMapping("/activation-status")
class ActivationStatusController(
    private val paymentConfig: PaymentConfigProperties,
    private val newPosConfig: PayUNewPosConfigProperties,
    private val transactionRepository: TransactionRepository,
) {
    private val log = LoggerFactory.getLogger(ActivationStatusController::class.java)
    private val objectMapper = ObjectMapper()
    private val tokenCache = PayUOAuthTokenCache()
    private val restClient = RestClient.builder().baseUrl(PRODUCTION_BASE_URL).build()

    @GetMapping
    fun show(@RequestParam token: String?): ResponseEntity<Any> {
        if (!MessageDigest.isEqual(paymentConfig.activationCheckToken.toByteArray(), (token ?: "").toByteArray())) {
            return ResponseEntity.status(401).body(mapOf("error" to "unauthorized"))
        }

        val newPos = fetchNewPosStatus()
        val since24h = Instant.now().minus(1, ChronoUnit.DAYS)

        return ResponseEntity.ok(
            ActivationStatusResponse(
                checkedAt = Instant.now().toString(),
                newPos = newPos,
                paidLast24h = transactionRepository.countByStatusAndPaidAtAfter(TransactionStatus.PAID, since24h),
                pendingLast24h = transactionRepository.countByStatusAndCreatedAtAfter(TransactionStatus.PENDING, since24h),
                activationLikelyDone = (newPos.methods ?: 0) > 0,
            ),
        )
    }

    private fun fetchNewPosStatus(): NewPosStatus {
        val token = try {
            tokenCache.getOrFetch(PayUOAuthTokenCache.NEW_POS_KEY) { fetchAccessToken() }
        } catch (e: Exception) {
            log.warn("ActivationStatus: OAuth PayU (newpos) nieudane: {}", e.message)
            return NewPosStatus(error = "oauth_failed")
        }

        return try {
            val responseText = restClient.get()
                .uri("/api/v2_1/paymethods?lang=pl")
                .headers { it.setBearerAuth(token) }
                .retrieve()
                .body(String::class.java) ?: "{}"

            val enabled = objectMapper.readTree(responseText).path("payByLinks")
                .filter { it.path("status").asText() == "ENABLED" }

            NewPosStatus(
                methods = enabled.size,
                blikEnabled = enabled.any { it.path("value").asText() == "blik" },
            )
        } catch (e: Exception) {
            log.warn("ActivationStatus: pobranie paymethods (newpos) nieudane: {}", e.message)
            NewPosStatus(error = "exception")
        }
    }

    private fun fetchAccessToken(): Pair<String, Long> {
        val form = "grant_type=client_credentials" +
            "&client_id=${newPosConfig.clientId}" +
            "&client_secret=${newPosConfig.clientSecret}"

        val bodyText = restClient.post()
            .uri("/pl/standard/user/oauth/authorize")
            .contentType(MediaType.APPLICATION_FORM_URLENCODED)
            .body(form)
            .retrieve()
            .body(String::class.java) ?: throw IllegalStateException("Brak odpowiedzi z endpointu OAuth PayU (newpos)")

        val json = objectMapper.readTree(bodyText)
        val token = json.path("access_token").takeIf { it.isTextual }?.asText()
            ?: throw IllegalStateException("Odpowiedź OAuth PayU (newpos) bez access_token")
        val expiresIn = json.path("expires_in").takeIf { it.isIntegralNumber }?.asLong() ?: 43199L
        return token to expiresIn
    }

    companion object {
        private const val PRODUCTION_BASE_URL = "https://secure.payu.com"
    }
}
